// mdb-cli/commands/info-command.js
'use strict';

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var Command = require('./command');

var Receiver = require('../receivers/receiver');

var BackendReceiver = require('../receivers/backend-receiver');

var DatabaseReceiver = require('../receivers/database-receiver');

var WordpressReceiver = require('../receivers/wordpress-receiver');

var Entity = require('../models/entity');

var InfoCommand = /*#__PURE__*/function (_Command) {
  _inherits(InfoCommand, _Command);

  var _super = _createSuper(InfoCommand);

  function InfoCommand(context) {
    var _this;

    _classCallCheck(this, InfoCommand);

    _this = _super.call(this, context);
    _this.receiver = null;
    _this.context = context;
    return _this;
  }

  _createClass(InfoCommand, [{
    key: "execute",
    value: function () {
      var _execute = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var _this2 = this;

        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return this.setReceiver();

              case 2:
                if (!this.receiver) {
                  _context.next = 11;
                  break;
                }

                if (!this.receiver.flags.help) {
                  _context.next = 5;
                  break;
                }

                return _context.abrupt("return", this.help());

              case 5:
                this.receiver.result.on('mdb.cli.live.output', function (msg) {
                  return _this2.printResult([msg]);
                });
                _context.next = 8;
                return this.receiver.info();

              case 8:
                this.printResult([this.receiver.result]);
                _context.next = 19;
                break;

              case 11:
                _context.next = 13;
                return this.detectReceiver();

              case 13:
                if (this.receiver) {
                  _context.next = 15;
                  break;
                }

                return _context.abrupt("return", this.help());

              case 15:
                this.receiver.result.on('mdb.cli.live.output', function (msg) {
                  return _this2.printResult([msg]);
                });
                _context.next = 18;
                return this.receiver.info();

              case 18:
                this.printResult([this.receiver.result]);

              case 19:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function execute() {
        return _execute.apply(this, arguments);
      }

      return execute;
    }()
  }, {
    key: "setReceiver",
    value: function () {
      var _setReceiver = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                _context2.t0 = this.entity;
                _context2.next = _context2.t0 === Entity.Backend ? 3 : _context2.t0 === Entity.Database ? 5 : _context2.t0 === Entity.Wordpress ? 7 : 9;
                break;

              case 3:
                this.receiver = new BackendReceiver(this.context);
                return _context2.abrupt("break", 10);

              case 5:
                this.receiver = new DatabaseReceiver(this.context);
                return _context2.abrupt("break", 10);

              case 7:
                this.receiver = new WordpressReceiver(this.context);
                return _context2.abrupt("break", 10);

              case 9:
                return _context2.abrupt("break", 10);

              case 10:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function setReceiver() {
        return _setReceiver.apply(this, arguments);
      }

      return setReceiver;
    }()
  }, {
    key: "detectReceiver",
    value: function () {
      var _detectReceiver = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee3() {
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return Receiver.detectEntity(this.context);

              case 2:
                this.entity = _context3.sent;
                _context3.next = 5;
                return this.setReceiver();

              case 5:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      function detectReceiver() {
        return _detectReceiver.apply(this, arguments);
      }

      return detectReceiver;
    }()
  }, {
    key: "help",
    value: function help() {
      this.results.addTextLine('Displays info about entity (current status).');
      this.results.addTextLine('\nUsage: mdb [entity] info [args]');
      this.results.addTextLine('\nAvailable entities: database, backend, wordpress');
      this.results.addTextLine('\nFlags:');
      this.results.addTextLine('  -n, --name \tProject name');
      this.printResult([this.results]);
    }
  }]);

  return InfoCommand;
}(Command);

module.exports = InfoCommand;
