// mdb-cli/commands/ls-command.js
'use strict';

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

var Command = require('./command');

var StarterReceiver = require('../receivers/starter-receiver');

var FrontendReceiver = require('../receivers/frontend-receiver');

var BackendReceiver = require('../receivers/backend-receiver');

var WordpressReceiver = require('../receivers/wordpress-receiver');

var DatabaseReceiver = require('../receivers/database-receiver');

var OrderReceiver = require('../receivers/order-receiver');

var Entity = require('../models/entity');

var LsCommand = /*#__PURE__*/function (_Command) {
  _inherits(LsCommand, _Command);

  var _super = _createSuper(LsCommand);

  function LsCommand(context) {
    var _this;

    _classCallCheck(this, LsCommand);

    _this = _super.call(this, context);
    _this.receiver = null;
    _this.context = context;
    _this.starterReceiver = null;
    _this.frontendReceiver = null;
    _this.backendReceiver = null;
    _this.wordpressReceiver = null;
    _this.databaseReceiver = null;
    _this.orderReceiver = null;

    _this.checkFlags(context);

    _this.setReceiver(context);

    return _this;
  }

  _createClass(LsCommand, [{
    key: "execute",
    value: function () {
      var _execute = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
        var _this2 = this;

        var flags;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                flags = this.context.getParsedFlags();

                if (!flags.help) {
                  _context.next = 3;
                  break;
                }

                return _context.abrupt("return", this.help());

              case 3:
                if (!this.receiver) {
                  _context.next = 10;
                  break;
                }

                this.receiver.result.on('mdb.cli.live.output', function (msg) {
                  _this2.printResult([msg]);
                });
                _context.next = 7;
                return this.receiver.list();

              case 7:
                this.printResult([this.receiver.result]);
                _context.next = 30;
                break;

              case 10:
                ['starterReceiver', 'frontendReceiver', 'backendReceiver', 'wordpressReceiver', 'databaseReceiver', 'orderReceiver'].forEach(function (receiver) {
                  _this2[receiver].result.on('mdb.cli.live.output', function (msg) {
                    _this2.printResult([msg]);
                  });
                });
                this.starterReceiver.result.addTextLine('\nStarters:');
                _context.next = 14;
                return this.starterReceiver.list();

              case 14:
                this.frontendReceiver.result.addTextLine('\nFrontend projects:');
                _context.next = 17;
                return this.frontendReceiver.list();

              case 17:
                this.backendReceiver.result.addTextLine('\nBackend projects:');
                _context.next = 20;
                return this.backendReceiver.list();

              case 20:
                this.backendReceiver.result.addTextLine('\nWordPress projects:');
                _context.next = 23;
                return this.wordpressReceiver.list();

              case 23:
                this.databaseReceiver.result.addTextLine('\nDatabases:');
                _context.next = 26;
                return this.databaseReceiver.list();

              case 26:
                this.orderReceiver.result.addTextLine('\nOrders:');
                _context.next = 29;
                return this.orderReceiver.list();

              case 29:
                this.printResult([this.frontendReceiver.result, this.backendReceiver.result, this.wordpressReceiver.result, this.databaseReceiver.result, this.starterReceiver.result, this.orderReceiver.result]);

              case 30:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function execute() {
        return _execute.apply(this, arguments);
      }

      return execute;
    }()
  }, {
    key: "checkFlags",
    value: function checkFlags(ctx) {
      if (this.entity === '') {
        var flags = ctx.getParsedFlags();
        if (!flags.all) this.entity = 'frontend';
      }
    }
  }, {
    key: "setReceiver",
    value: function setReceiver(ctx) {
      switch (this.entity) {
        case Entity.Starter:
          this.receiver = new StarterReceiver(ctx);
          break;

        case Entity.Frontend:
          this.receiver = new FrontendReceiver(ctx);
          break;

        case Entity.Backend:
          this.receiver = new BackendReceiver(ctx);
          break;

        case Entity.Wordpress:
          this.receiver = new WordpressReceiver(ctx);
          break;

        case Entity.Database:
          this.receiver = new DatabaseReceiver(ctx);
          break;

        case Entity.Order:
          this.receiver = new OrderReceiver(ctx);
          break;

        default:
          this.starterReceiver = new StarterReceiver(ctx);
          this.frontendReceiver = new FrontendReceiver(ctx);
          this.backendReceiver = new BackendReceiver(ctx);
          this.wordpressReceiver = new WordpressReceiver(ctx);
          this.databaseReceiver = new DatabaseReceiver(ctx);
          this.orderReceiver = new OrderReceiver(ctx);
          break;
      }
    }
  }, {
    key: "help",
    value: function help() {
      this.results.addTextLine('List entity content.');
      this.results.addTextLine('\nUsage: mdb [entity] ls');
      this.results.addTextLine('\nAvailable entities: starter, frontend (default), backend, wordpress, database, order');
      this.results.addTextLine('\nFlags:');
      this.results.addTextLine('  -a, --all  \tList all starters, projects, databases and orders.');
      this.results.addTextLine('  -o, --only \tSpecify the type of starters you want to display. Possible values: frontend, backend, wordpress');
      this.printResult([this.results]);
    }
  }]);

  return LsCommand;
}(Command);

module.exports = LsCommand;
