// mdb-cli/helpers/download-from-ftp.js
'use strict';

var ProgressBar = require('progress');

var _require = require('stream'),
    Readable = _require.Readable;

var unzip = require('unzipper');

module.exports = {
  downloadFromFTP: function downloadFromFTP(http, options, destination) {
    return new Promise(function (resolve, reject) {
      var request = http.createRawRequest(options, function (response) {
        var result,
            message = '';
        var readStream = new Readable();

        readStream._read = function () {};

        var len = Number(response.headers['content-length']);
        var bar = new ProgressBar('[:bar] :eta s', {
          complete: '=',
          incomplete: ' ',
          width: 100,
          total: len
        });
        response.on('data', function (chunk) {
          if (response.statusCode === 200) {
            readStream.push(chunk);
            bar.tick(chunk.length);
          } else {
            message += Buffer.from(chunk).toString('utf8');
          }
        });
        response.on('end', function () {
          var statusCode = response.statusCode,
              statusMessage = response.statusMessage;

          if (statusCode === 200) {
            result = 'Download completed.';
            readStream.push(null);
          } else {
            reject(message || statusMessage);
          }
        });

        try {
          readStream.pipe(unzip.Extract({
            path: destination
          })).on('close', function () {
            return resolve(result);
          });
        } catch (e) {
          reject('Download error.');
        }
      });
      request.on('error', reject);
      request.end();
    });
  }
};
