// mdb-cli/utils/output-printer.js
'use strict';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var OutputPrinter = /*#__PURE__*/function () {
  function OutputPrinter() {
    _classCallCheck(this, OutputPrinter);

    this.consoleColors = {
      red: '\x1b[31m%s\x1b[0m',
      green: '\x1b[32m%s\x1b[0m',
      yellow: '\x1b[33m%s\x1b[0m',
      blue: '\x1b[34m%s\x1b[0m',
      purple: '\x1b[35m%s\x1b[0m',
      turquoise: '\x1b[36m%s\x1b[0m',
      grey: '\x1b[37m%s\x1b[0m',
      inverted: '\x1b[40m%s\x1b[0m',
      invertedRed: '\x1b[41m%s\x1b[0m',
      invertedGreen: '\x1b[42m%s\x1b[0m',
      invertedYellow: '\x1b[43m%s\x1b[0m',
      invertedBlue: '\x1b[44m%s\x1b[0m',
      invertedPurple: '\x1b[45m%s\x1b[0m',
      invertedTurquoise: '\x1b[46m%s\x1b[0m',
      invertedGrey: '\x1b[47m%s\x1b[0m'
    };
  }
  /**
   * @param results: CommandResult[]
   */


  _createClass(OutputPrinter, [{
    key: "print",
    value: function print(results) {
      var _this = this;

      results.forEach(function (cr) {
        cr.messages.forEach(function (m) {
          if (m.type === 'text') {
            console.log(m.value);
          } else if (m.type === 'table') {
            console.table(m.value);
          } else if (m.type === 'alert') {
            console.log(_this.consoleColors[m.color], m.value.title, m.value.body);
          } else {
            console.log(m.value);
          }
        });
      });
    }
  }]);

  return OutputPrinter;
}();

module.exports = OutputPrinter;
